pkg/agent/controller/trafficcontrol/controller.go

```diff
@@ -0,0 +1,1020 @@
+	return portUUID, err
+}
+
+func (c *Controller) createGREPort(portName string, tunnelConfig *v1alpha2.GRETunnel) (string, error) {
+	extraOptions := map[string]interface{}{}
+	if tunnelConfig.Key != nil {
+		extraOptions["key"] = strconv.Itoa(int(*tunnelConfig.Key))
+	}
+	portUUID, err := c.ovsBridgeClient.CreateTunnelPortExt(portName,
+		ovsconfig.GRETunnel,
+		0,
+		false,
+		"",
+		tunnelConfig.RemoteIP,
+		"",
+		extraOptions,
+		trafficControlPortExternalIDs)
+	return portUUID, err
+}
+
+func (c *Controller) createERSPANPort(portName string, tunnelConfig *v1alpha2.ERSPANTunnel) (string, error) {
+	extraOptions := make(map[string]interface{})
+	extraOptions["erspan_ver"] = strconv.Itoa(int(tunnelConfig.Version))
+	if tunnelConfig.SessionID != nil {
+		extraOptions["key"] = strconv.Itoa(int(*tunnelConfig.SessionID))
+	}
+	if tunnelConfig.Version == 1 {
+		if tunnelConfig.Index != nil {
+			extraOptions["erspan_idx"] = strconv.FormatInt(int64(*tunnelConfig.Index), 16)
+		}
+	} else if tunnelConfig.Version == 2 {
+		if tunnelConfig.Dir != nil {
+			extraOptions["erspan_dir"] = strconv.Itoa(int(*tunnelConfig.Dir))
+		}
+		if tunnelConfig.HardwareID != nil {
+			extraOptions["erspan_hwid"] = strconv.Itoa(int(*tunnelConfig.HardwareID))
+		}
+	}
+	portUUID, err := c.ovsBridgeClient.CreateTunnelPortExt(portName,
+		ovsconfig.ERSPANTunnel,
+		0,
+		false,
+		"",
+		tunnelConfig.RemoteIP,
+		"",
+		extraOptions,
+		trafficControlPortExternalIDs)
+	return portUUID, err
+}
+
+func (c *Controller) getPortName(port *v1alpha2.TrafficControlPort) string {
+	var portName string
+	switch {
+	case port.OVSInternal != nil:
+		portName = port.OVSInternal.Name
+	case port.Device != nil:
+		portName = port.Device.Name
+	case port.VXLAN != nil:
+		portName = genVXLANPortName(port.VXLAN)
+	case port.GENEVE != nil:
+		portName = genGENEVEPortName(port.GENEVE)
+	case port.GRE != nil:
+		portName = genGREPortName(port.GRE)
+	case port.ERSPAN != nil:
+		portName = genERSPANPortName(port.ERSPAN)
+	}
+	return portName
+}
+
+// getOrCreateTrafficControlPort ensures that there is an OVS port for the given TrafficControlPort and binds the port
+// to the TrafficControl. The OVS port will be created if the port doesn't exist. It returns the ofPort of the OVS port
+// on success, a boolean indicating whether the port exists, and an error if there is.
+func (c *Controller) getOrCreateTrafficControlPort(port *v1alpha2.TrafficControlPort, portName, tcName string) (uint32, bool, error) {
+	c.ovsPortUpdateMutex.Lock()
+	defer c.ovsPortUpdateMutex.Unlock()
+
+	if itf, ok := c.interfaceStore.GetInterfaceByName(portName); ok {
+		// Bind the port to the TrafficControl.
+		if _, exists := c.portToTCBindings[portName]; !exists {
+			c.portToTCBindings[portName] = &portToTCBinding{
+				ofPort:          uint32(itf.OFPort),
+				trafficControls: sets.NewString(),
+			}
+		}
+		c.portToTCBindings[portName].trafficControls.Insert(tcName)
+		return uint32(itf.OFPort), true, nil
```

Comment:
I think you need to return false and update the method's comment as you use this bool as `portCreated` on caller side. "portCreated" should mean whether the port is created by the current invocation, like the meaning of "portDeleted"

Reply:
Not needed anymore.